### PR TITLE
Mem percent watermark

### PR DESCRIPTION
**Pull request by chenkaiyue**
watermark of memory usage percent

### CODE REVIEW

**Comment by chenkaiyue** [comment]
1. deploy 2 pods which use 3.5G and 2G memory

2.  node use 78.7% memory  
企业微信截图_cd6e5d90-8a7b-4b3f-b213-b51de7afaa01

3. set watermark of memory usage percent to 50%

5. pod with 3.5G memory has been evicted, pod with 2G memory is still running
企业微信截图_16995c41-fbd0-45c3-b968-63e5eeb35a71

**Comment by chenkaiyue** [comment]
@mfanjie  We can take this pr as standard, this pr includes all 3 features.

File: pkg/ensurance/executor/mem_usage.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	EvictQuantified: true,
+	EvictFunc:       memUsageEvictPod,
+}
+
+func memUsageEvictPod(wg *sync.WaitGroup, ctx *ExecuteContext, index int, totalReleasedResource *ReleaseResource, EvictPods EvictPods) (errPodKeys []string, released ReleaseResource) {
+	wg.Add(1)
+
+	// Calculate release resources
+	released = releaseMemUsage(EvictPods[index])
+	totalReleasedResource.Add(released)
```
Comment (yan234280533) [inline pkg/ensurance/executor/mem_usage.go:39]: If the pod evicted failed, need we subtract  the release?

File: pkg/ensurance/executor/sort/mem_usage_sort.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package sort
+
+import (
+	"github.com/gocrane/crane/pkg/ensurance/executor/podinfo"
+	"github.com/gocrane/crane/pkg/utils"
+)
+
+func MemUsageSort(pods []podinfo.PodContext) {
```
Comment (yan234280533) [inline pkg/ensurance/executor/sort/mem_usage_sort.go:8]: Can we add some tests for memory sort ?

File: examples/ensurance/evict-on-mem-usage-percent/watermark.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    nodeLocalGet:
+      localCacheTTLSeconds: 60
+    timeoutSeconds: 10
+  rules:
+    - actionName: eviction
+      avoidanceThreshold: 2
+      metricRule:
+        name: memory_total_utilization
+        value: 50
+      name: cpu-usage-percent
```
Comment (yan234280533) [inline examples/ensurance/evict-on-mem-usage-percent/watermark.yaml:16]: The name need modified to "mem-usage-percent"

File: examples/ensurance/evict-on-mem-usage-total/pod.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        - "2"
+        - --vm-bytes
+        - "2G"
+      name: stress
+      volumeMounts:
+        - mountPath: /data
+          name: data
+  volumes:
+    - hostPath:
+        path: /data/dd
```
Comment (yan234280533) [inline examples/ensurance/evict-on-mem-usage-total/pod.yaml:25]: Why we added a path to the container? It is necessary

File: examples/ensurance/evict-on-mem-usage-total/watermark.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    nodeLocalGet:
+      localCacheTTLSeconds: 60
+    timeoutSeconds: 10
+  rules:
+    - actionName: eviction
+      avoidanceThreshold: 2
+      metricRule:
+        name: memory_total_usage
+        value: 5000000000
+      name: cpu-usage-percent
```
Comment (yan234280533) [inline examples/ensurance/evict-on-mem-usage-total/watermark.yaml:16]: The name need modified to "mem-usage

File: examples/ensurance/evict-on-mem-usage-total/pod.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: low-pi6
```
Comment (yan234280533) [inline examples/ensurance/evict-on-mem-usage-total/pod.yaml:4]: Is "low-pi1"  more better  here?

File: examples/ensurance/evict-on-mem-usage-percent/pod.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: low-pi6
```
Comment (yan234280533) [inline examples/ensurance/evict-on-mem-usage-percent/pod.yaml:4]: Is "low-pi1" more better here?

File: examples/ensurance/evict-on-mem-usage-total/watermark.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  nodeQualityProbe:
+    nodeLocalGet:
+      localCacheTTLSeconds: 60
+    timeoutSeconds: 10
+  rules:
+    - actionName: eviction
+      avoidanceThreshold: 2
+      metricRule:
+        name: memory_total_usage
+        value: 5000000000
```
Comment (yan234280533) [inline examples/ensurance/evict-on-mem-usage-total/watermark.yaml:15]: Can it add an explanatory note this is "5Gi". Users can not know the unit conversion size

File: examples/ensurance/evict-on-mem-usage-percent/pod.yaml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  labels:
+    k8s-app: low
+    preemptible_job: "true"
+spec:
+  containers:
+    - image: ccr.ccs.tencentyun.com/tkeimages/stress-ng:v0.12.09
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply examples/ensurance/evict-on-mem-usage-percent/pod.yaml:10]: ok

File: examples/ensurance/evict-on-mem-usage-total/elastic-pod-qos.yaml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+apiVersion: ensurance.crane.io/v1alpha1
+kind: PodQOS
+metadata:
+  name: all-elastic-pods
+spec:
+  allowedActions:
+    - eviction
+  resourceQOS:
+    cpuQOS:
+      cpuPriority: 7
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply examples/ensurance/evict-on-mem-usage-total/elastic-pod-qos.yaml:10]: I have deleted it

File: pkg/ensurance/executor/mem_usage.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+			return
+		}
+		klog.Warningf("Evicting pod %v", evictPod.Key)
+		err = utils.EvictPodWithGracePeriod(ctx.Client, pod, evictPod.DeletionGracePeriodSeconds)
+		if err != nil {
+			errPodKeys = append(errPodKeys, "evict failed ", evictPod.Key.String())
+			klog.Warningf("Failed to evict pod %s: %v", evictPod.Key.String(), err)
+			return
+		}
+		metrics.ExecutorEvictCountsInc()
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/executor/mem_usage.go:56]: ok

File: pkg/ensurance/executor/mem_usage.go (hunk 1 above)
Comment (chenkaiyue) [reply pkg/ensurance/executor/mem_usage.go:39]: Ok, i will make a new pr to fix this.

File: examples/ensurance/evict-on-mem-usage-percent/watermark.yaml (hunk 3 above)
Comment (chenkaiyue) [reply examples/ensurance/evict-on-mem-usage-percent/watermark.yaml:16]: ok

File: examples/ensurance/evict-on-mem-usage-total/pod.yaml (hunk 4 above)
Comment (chenkaiyue) [reply examples/ensurance/evict-on-mem-usage-total/pod.yaml:25]: ok

File: pkg/ensurance/executor/sort/mem_usage_sort.go (hunk 2 above)
Comment (chenkaiyue) [reply pkg/ensurance/executor/sort/mem_usage_sort.go:8]: done

**Comment by chenkaiyue** [comment]
Mem usage test:
https://github.com/gocrane/crane/pull/554#issuecomment-1247721045

**Comment by chenkaiyue** [comment]
Mem percent test:
https://github.com/gocrane/crane/pull/557#issuecomment-1247924500

**Review by yan234280533 — APPROVED** [review]

**Pull request merged**